syntax = "proto3";

import "envoy/service/tap/v2alpha/common.proto";
import "envoy/api/v2/core/config_source.proto";

import "validate/validate.proto";

package envoy.config.common.tap.v2alpha;

option java_outer_classname = "CommonProto";
option java_multiple_files = true;
option java_package = "io.envoyproxy.envoy.config.common.tap.v2alpha";

// [#protodoc-title: Common tap extension configuration]

// Common configuration for all tap extensions.
message CommonExtensionConfig {

  message TDSConfig {
    // Configuration for the source of EDS updates for this Cluster.
    envoy.api.v2.core.ConfigSource config_source = 1;

    // Tap config to request from XDS server.
    string name = 2;
  }

  oneof config_type {
    option (validate.required) = true;

    // If specified, the tap filter will be configured via an admin handler.
    AdminConfig admin_config = 1;

    // If specified, the tap filter will be configured via a static configuration that cannot be
    // changed.
    service.tap.v2alpha.TapConfig static_config = 2;

    // Configuration to use for TDS updates for the filter.
    TDSConfig tds_config = 3;
  }
}

// Configuration for the admin handler. See :ref:`here <config_http_filters_tap_admin_handler>` for
// more information.
message AdminConfig {
  // Opaque configuration ID. When requests are made to the admin handler, the passed opaque ID is
  // matched to the configured filter opaque ID to determine which filter to configure.
  string config_id = 1 [(validate.rules).string.min_bytes = 1];
}
